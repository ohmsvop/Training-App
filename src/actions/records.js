import database from '../firebase/firebase';

export const addRecord = (record) => ({
  type: 'ADD_RECORD',
  record
});

export const startAddRecord = (recordData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      detail = '',
      createdAt = 0
    } = recordData;
    const record = { title, detail, createdAt }
    return database.ref(`users/${uid}/records`).push(record).then((ref) => {
      dispatch(addRecord({
        id: ref.key,
        ...record
      }));
    });
  };
};