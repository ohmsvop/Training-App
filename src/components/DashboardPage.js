import React from 'react';
import { Link } from 'react-router-dom';

const DashboardPage = () => (
  <div>
    Dashboard page content
    <Link to="/create">Create</Link>
  </div>
);

export default DashboardPage;
