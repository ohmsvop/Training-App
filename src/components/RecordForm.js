import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class RecordForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      movement: props.record ? props.record.movement : '',
      note: props.record ? props.record.note : '',
      weight: props.record ? props.record.weight : '',
      rep: props.record ? props.record.rep : '',
      set: props.record ? props.record.set : '',
      createdAt: props.record ? moment(props.record.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onMovementChange = (e) => {
    const movement = e.target.value;
    this.setState(() => ({ movement }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onWeightChange = (e) => {
    const weight = e.target.value;
    if (!weight || weight.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ weight }));
    }
  };
  onRepChange = (e) => {
    const rep = e.target.value;
    if (!rep || rep.match(/^\d+$/)) {
      this.setState(() => ({ rep }));
    }
  };
  onSetChange = (e) => {
    const set = e.target.value;
    if (!set || set.match(/^\d+$/)) {
      this.setState(() => ({ set }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.movement || !this.state.weight || !this.state.rep || !this.state.set) {
      this.setState(() => ({ error: 'Please provide movement, weight, rep, and set' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        movement: this.state.movement,
        weight: this.state.weight,
        rep: this.state.rep,
        set: this.state.set,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}

        <input
          type="text"
          placeholder="Movement"
          autoFocus
          className="text-input"
          value={this.state.movement}
          onChange={this.onMovementChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Weight"
          value={this.state.weight}
          onChange={this.onWeightChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Rep"
          value={this.state.rep}
          onChange={this.onRepChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Set"
          value={this.state.set}
          onChange={this.onSetChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your record (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <div>
          <button className="button">Save Record</button>
        </div>
      </form>
    )
  }
};