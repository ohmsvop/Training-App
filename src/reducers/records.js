const recordsReducerDefaultState = [];

const recordsReducer = (state = recordsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_RECORD':
      return [
        ...state,
        action.record
      ];
    // case 'REMOVE_RECORD':
    //   return state.filter(({ id }) => id !== action.id);
    // case 'EDIT_RECORD':
    //   return state.map((record) => {
    //     if (record.id === action.id) {
    //       return {
    //         ...record,
    //         ...action.updates
    //       };
    //     } else {
    //       return record;
    //     }
    //   });
    // case 'SET_RECORDS':
    //   return action.records;
    default:
      return state;
  }
};

export default recordsReducer;