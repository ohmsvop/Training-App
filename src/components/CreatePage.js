import React from 'react';
import { connect } from 'react-redux';
import RecordForm from './RecordForm';
import { startAddRecord } from '../actions/records';

export class AddRecordPage extends React.Component {
  onSubmit = (record) => {
    this.props.startAddRecord(record)
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Record</h1>
          </div>
        </div>
        <div className="content-container">
          <RecordForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = (dispatch) => ({
  startAddRecord: (record) => dispatch(startAddRecord(record))
});

export default connect(undefined, mapDispatchToProps)(AddRecordPage);
