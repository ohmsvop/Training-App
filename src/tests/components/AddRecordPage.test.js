import React from 'react';
import { shallow } from 'enzyme';
import { AddRecordPage } from '../../components/AddRecordPage';
import records from '../fixtures/records';

let startAddRecord, history, wrapper;

beforeEach(() => {
  startAddRecord = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddRecordPage startAddRecord={startAddRecord} history={history} />);
});

test('should render AddRecordPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('RecordForm').prop('onSubmit')(records[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startAddRecord).toHaveBeenLastCalledWith(records[1]);
});