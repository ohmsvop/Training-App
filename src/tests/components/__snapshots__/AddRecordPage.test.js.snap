// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddRecordPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Record
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <RecordForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
